packages/signature-controller/src/SignatureController.ts

```diff
@@ -0,0 +1,678 @@
+  unapprovedPersonalMsgCount: number;
+  unapprovedTypedMessagesCount: number;
+};
+
+type AllowedActions = AddApprovalRequest | AcceptRequest | RejectRequest;
+
+export type GetSignatureState = {
+  type: `${typeof controllerName}:getState`;
+  handler: () => SignatureControllerState;
+};
+
+export type SignatureStateChange = {
+  type: `${typeof controllerName}:stateChange`;
+  payload: [SignatureControllerState, Patch[]];
+};
+
+export type SignatureControllerActions = GetSignatureState;
+
+export type SignatureControllerEvents = SignatureStateChange;
+
+export type SignatureControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  SignatureControllerActions | AllowedActions,
+  SignatureControllerEvents,
+  AllowedActions['type'],
+  never
+>;
+
+export interface KeyringController {
+  signMessage: (messsageParams: MessageParams) => Promise<string>;
+  signPersonalMessage: (
+    messsageParams: PersonalMessageParams,
+  ) => Promise<string>;
+  signTypedMessage: (
+    messsageParams: TypedMessageParams,
+    options: { version: string | undefined },
+  ) => Promise<string>;
+}
+
+export type SignatureControllerOptions = {
+  messenger: SignatureControllerMessenger;
+  keyringController: KeyringController;
+  isEthSignEnabled: () => boolean;
+  getState: () => any;
+  securityProviderRequest?: (
+    requestData: any,
+    methodName: string,
+  ) => Promise<any>;
+};
+
+/**
+ * Controller for creating signing requests requiring user approval.
+ */
+export class SignatureController extends BaseControllerV2<
+  typeof controllerName,
+  SignatureControllerState,
+  SignatureControllerMessenger
+> {
+  hub: EventEmitter;
+
+  private _keyringController: KeyringController;
+
+  private _isEthSignEnabled: () => boolean;
+
+  private _getState: () => any;
+
+  private _messageManager: MessageManager;
+
+  private _personalMessageManager: PersonalMessageManager;
+
+  private _typedMessageManager: TypedMessageManager;
+
+  /**
+   * Construct a Sign controller.
+   *
+   * @param options - The controller options.
+   * @param options.messenger - The restricted controller messenger for the sign controller.
+   * @param options.keyringController - An instance of a keyring controller used to perform the signing operations.
+   * @param options.isEthSignEnabled - Callback to return true if eth_sign is enabled.
+   * @param options.getState - Callback to retrieve all user state.
+   * @param options.securityProviderRequest - A function for verifying a message, whether it is malicious or not.
+   */
+  constructor({
+    messenger,
+    keyringController,
+    isEthSignEnabled,
+    getState,
+    securityProviderRequest,
+  }: SignatureControllerOptions) {
+    super({
+      name: controllerName,
+      metadata: stateMetadata,
+      messenger,
+      state: getDefaultState(),
+    });
+
+    this._keyringController = keyringController;
+    this._isEthSignEnabled = isEthSignEnabled;
+    this._getState = getState;
```

Comment:
Since the controller has state, I wonder if consumers could get confused into thinking that this has to do with the controller. Whose state is this? Should this option be renamed to something clearer?

Reply:
Done, renamed to `getAllState`.